fars/booking/gcal.py

```diff
@@ -0,0 +1,215 @@
+            body['location'] = booking.bookable.google_calendar_event_location
+        if 'visibility' not in body:
+            body['visibility'] = 'private'
+        if 'conferenceData' not in body:
+            body['conferenceData'] = {
+                'createRequest': {
+                    'conferenceSolutionKey': {
+                        'type': 'hangoutsMeet',
+                    },
+                },
+            },
+        if 'organizer' not in body and booking.user.email:
+            body['organizer'] = {
+                'email': booking.user.email,
+            },
+        if 'recurrence' not in body and booking.recurrence:
+            f = booking.recurrence['frequency']
+            date_str = booking.recurrence['repeat_until'].replace('-', '')
+            body['recurrence'] = [
+                f'RRULE:FREQ=DAILY;INTERVAL={f};UNTIL={date_str}T235959Z',
+            ]
+
+        return body
+
+    def try_create_event(self, booking):
+        try:
+            return gcal_service.events().insert(
+                calendarId=self.calendar_id,
+                body=self._create_event_body(booking),
+            ).execute()
+
+        except HttpError as e:
+            logger.error(e)
+
+        return None
+
+    def try_update_event(self, booking):
+        try:
+            event = self.try_get_event(booking)
+
+            return gcal_service.events().update(
+                calendarId=self.calendar_id,
+                eventId=booking.google_calendar_event_id,
+                body=self._create_event_body(booking, event),
+            ).execute()
+
+        except HttpError as e:
+            logger.error(e)
+
+        return None
+
+    """
+    Get the Google Calendar event that corresponds to a certain booking.
+    """
+    def try_get_event(self, booking):
+        try:
+            event = gcal_service.events().get(
+                calendarId=self.calendar_id,
+                eventId=booking.google_calendar_event_id,
+            ).execute()
+            return event
+
+        except HttpError as e:
+            logger.error(e)
+
+        return None
+
+    '''
+    Delete the Google Calendar event for a booking, if it exist.
+    '''
+    def try_delete_event(self, booking):
+        try:
+            gcal_service.events().delete(
+                calendarId=self.calendar_id,
+                eventId=booking.google_calendar_event_id
+            ).execute()
+            return True
+
+        except HttpError as e:
+            error = json.loads(e.content)['error'] # = { errors, code, message }
+
+            # If the event has already been deleted, OK
+            if error['code'] == 410:
+                return True
+            logger.error(e)
+
+        return False
+
+
+class GCalCreateEventThread(Thread):
+    def __init__(self, booking):
+        self.booking = booking
+        self.gcal = GoogleCalendar(booking.bookable.google_calendar_id)
+        Thread.__init__(self)
```

Comment:
IIRC threads in python are kind of a "fake" implementation. There is no actual parallelism, just async execution. I guess this is still fine though, as long as the request is allowed to finish handling and return the response without having to wait for the thread to finish.